### PR TITLE
Bug 1753672: Redeploy console deployment upon proxy changes

### PR DESCRIPTION
**Pull request by jhadvig**
The console deployment has not been rollout upon changes made to the Proxy CR.
+ refact a the Proxy codebase

Tested manually on latest cluster and it's fixing the issue.

/assign @benjaminapetersen 

FYI @spadgett

### CODE REVIEW

File: pkg/console/operator/operator.go [hunk 1]
```diff
@@ -75,8 +72,7 @@ type consoleOperator struct {
 	recorder       events.Recorder
 	resourceSyncer resourcesynccontroller.ResourceSyncer
 	//proxy
-	proxyCfgLister   configlisterv1.ProxyLister
-	proxyCfgInformer cache.SharedIndexInformer
+	proxy configclientv1.ProxyInterface
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: +1

File: pkg/console/operator/operator.go [hunk 2]
```diff
@@ -122,8 +117,7 @@ func NewConsoleOperator(
 		recorder:       recorder,
 		resourceSyncer: resourceSyncer,
 		// proxy
-		proxyCfgLister:   proxyCfgInformer.Lister(),
-		proxyCfgInformer: proxyCfgInformer.Informer(),
+		proxy: configClient.Proxies(),
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: nit: this is a config, so I may just put it at the top with the 3 existing configs, console, operator, infrastructure.

```go 
operatorConfigClient:       operatorConfigClient.Consoles(),
consoleConfigClient:        configClient.Consoles(),
infrastructureConfigClient: configClient.Infrastructures(),
proxyConfigClient: configClient.Proxies(),
```

File: pkg/console/subresource/deployment/deployment.go [hunk 3]
```diff
@@ -67,6 +69,7 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
 		configMapResourceVersionAnnotation:          cm.GetResourceVersion(),
 		serviceCAConfigMapResourceVersionAnnotation: serviceCAConfigMap.GetResourceVersion(),
 		trustedCAConfigMapResourceVersionAnnotation: trustedCAConfigMap.GetResourceVersion(),
+		proxyConfigResourceVersionAnnotation:        proxyConfig.GetResourceVersion(),
```
Comment (benjaminapetersen) [inline pkg/console/subresource/deployment/deployment.go]: This should force the pod rollout, yup.

File: pkg/console/operator/operator.go (hunk 1 above)
Comment (benjaminapetersen) [reply pkg/console/operator/operator.go]: We should really reorder here as well, 

```
type consoleOperator struct {
     // configs
    operatorConfigClient operatorclientv1.ConsoleInterface
    consoleConfigClient  configclientv1.ConsoleInterface
    infrastructureConfigClient configclientv1.InfrastructureInterface
    proxyConfigClient configclientv1.ProxyInterface
    //
```
to match below.

File: pkg/console/operator/operator.go [hunk 4]
```diff
@@ -108,6 +103,7 @@ func NewConsoleOperator(
 		operatorConfigClient:       operatorConfigClient.Consoles(),
 		consoleConfigClient:        configClient.Consoles(),
 		infrastructureConfigClient: configClient.Infrastructures(),
+		proxy:                      configClient.Proxies(),
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: `proxyConfigClient`? for consistency, since we named the others `*ConfigClient`

**Comment by jhadvig** [comment]
@benjaminapetersen comments addressed. PTAL

**Comment by benjaminapetersen** [comment]
/retest

`"Cluster did not complete upgrade: timed out waiting for the condition: Working towards 0.0.1-2019-09-19-182246: 17% complete"`

lots of things didn't make it.

**Comment by jhadvig** [comment]
```
2019/09/19 21:59:12 Container setup in pod e2e-aws-upgrade completed successfully
[INFO] Acquiring a lease ...
[INFO] Lease acquired!
[INFO] Leased resource: {"type":"aws-quota-slice","name":"5045341b-afbe-4b42-bf7d-28375d35e73a","state":"leased","owner":"ci-op-wi9ds52i-e9c6b","lastupdate":"2019-09-19T21:17:31.762975378Z","userdata":{}}
[INFO] Sending heartbeats to retain the lease...
heartbeat sent for resource "5045341b-afbe-4b42-bf7d-28375d35e73a"
heartbeat sent for resource "5045341b-afbe-4b42-bf7d-28375d35e73a"
...
Another process exited
[INFO] Releasing the lease on resouce 5045341b-afbe-4b42-bf7d-28375d35e73a...
failed to release resource "5045341b-afbe-4b42-bf7d-28375d35e73a": Post http://boskos.ci/release?dest=free&name=5045341b-afbe-4b42-bf7d-28375d35e73a&owner=ci-op-wi9ds52i-e9c6b: dial tcp 172.30.131.17:80: connect: no route to host
2019/09/19 23:08:37 Container lease in pod e2e-aws-upgrade failed, exit code 1, reason Error
2019/09/19 23:16:30 Copied 216.42Mi of artifacts from e2e-aws-upgrade to /logs/artifacts/e2e-aws-upgrade
2019/09/19 23:16:32 Ran for 2h5m41s
2019/09/19 23:16:32 Submitted failure event to sentry (id=3a381bfe05924ea0b260779f8ea2bb44)
error: could not run steps: step e2e-aws-upgrade failed: template pod "e2e-aws-upgrade" failed: the pod ci-op-wi9ds52i/e2e-aws-upgrade failed after 1h59m53s (failed containers: lease): ContainerFailed one or more containers exited
Container lease exited with code 1, reason Error
---
ce "5045341b-afbe-4b42-bf7d-28375d35e73a"
heartbeat sent for resource "5045341b-afbe-4b42-bf7d-28375d35e73a"
heartbeat sent for resource "5045341b-afbe-4b42-bf7d-28375d35e73a"
...
Another process exited
[INFO] Releasing the lease on resouce 5045341b-afbe-4b42-bf7d-28375d35e73a...
failed to release resource "5045341b-afbe-4b42-bf7d-28375d35e73a": Post http://boskos.ci/release?dest=free&name=5045341b-afbe-4b42-bf7d-28375d35e73a&owner=ci-op-wi9ds52i-e9c6b: dial tcp 172.30.131.17:80: connect: no route to host
---

**Comment by benjaminapetersen** [comment]
/retest

`could not wait for build: the build console-operator failed after 1m20s with reason PushImageToRegistryFailed: Failed to push the image to the registry.
`

**Pull request merged**